Make GCE load-balancers create health checks for nodes

MrHohn opened the pull request:
From #14661. Proposal on kubernetes/community#552. Fixes #46313.

Bullet points:
- Create nodes health check and firewall (for health checking) for non-OnlyLocal service.
- Create local traffic health check and firewall (for health checking) for OnlyLocal service.
- Version skew: 
   - Don't create nodes health check if any nodes has version < 1.7.0.
   - Don't backfill nodes health check on existing LBs unless users explicitly trigger it.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
GCE Cloud Provider: New created LoadBalancer type Service now have health checks for nodes by default.
An existing LoadBalancer will have health check attached to it when:
- Change Service.Spec.Type from LoadBalancer to others and flip it back.
- Any effective change on Service.Spec.ExternalTrafficPolicy.
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/45524)
<!-- Reviewable:end -->

MrHohn commented:
/assign @nicksardo 
FYI the new functionality is mostly implemented --- except the uid :)

nicksardo commented:
Cool, I spoke with Tim today about a complication with the UID. We don't have direct access to the config map from within the cloud provider. We need to decide on a solution tomorrow.

MrHohn commented:
@k8s-bot kops aws e2e test this

MrHohn commented:
@k8s-bot kops aws e2e test this, issue: #34676

MrHohn commented:
@k8s-bot kops aws e2e test this

nicksardo commented:
Fixes #40778

MrHohn commented on pkg/cloudprovider/providers/gce/gce_loadbalancer.go:
```diff
@@ -535,41 +548,77 @@ func (gce *GCECloud) deleteTargetPool(name, region string, hc *compute.HttpHealt
+			gce.nodesHealthCheckLock.Lock()
+			defer gce.nodesHealthCheckLock.Unlock()
+			// Ensure firewall for nodes health check.
+			if err := gce.ensureNodesHealthCheckFirewall("TODO-CLUSTERID", gce.region, hosts); err != nil {
```
I missed the firewall for only local traffic health check. Will update.

MrHohn replied:
Done.

MrHohn commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

MrHohn commented:
Rebased on top of #46147, removing WIP.

cc @bowei @thockin

MrHohn commented:
/unassign

MrHohn commented:
@k8s-bot pull-kubernetes-unit test this, issue: #42418

MrHohn commented:
Rebased, PR ready for review @nicksardo

MrHohn commented:
ACK https://github.com/kubernetes/kubernetes/pull/46147#issuecomment-303627556, working on it.

MrHohn commented:
> ACK https://github.com/kubernetes/kubernetes/pull/46147#issuecomment-303627556, working on it.

Done, except the async  part.

MrHohn commented:
@k8s-bot pull-kubernetes-node-e2e test this

nicksardo commented on pkg/cloudprovider/providers/gce/gce_loadbalancer.go:
```diff
@@ -875,6 +937,44 @@ func (gce *GCECloud) firewallNeedsUpdate(name, serviceName, region, ipAddress st
 
+func (gce *GCECloud) ensureHttpHealthCheckFirewall(region string, hosts []*gceInstance, hcName string, hcPort int32, isNodesHealthCheck bool) error {
+	// Prepare the firewall params for creating / checking.
+	desc := "Firewall for health checks from load-balancer."
```
```
if isNodesHealthCheck
    use [makeFirewallDescription](https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/gce/gce_loadbalancer.go#L835)
else 
    {thoughts here?}
```

`"Firewall for health checks from load-balancer."` and `{"kubernetes.io/service-name":"%s", "kubernetes.io/service-ip":"%s"}` are vastly different. I'm wondering if we should just leave the common case description empty.

MrHohn replied:
Yeah make sense to differentiate them. What about `{kubernetes.io/loadbalancer-nodes-healthcheck}`? I think anything meaningful here would help.

nicksardo commented on pkg/cloudprovider/providers/gce/gce_loadbalancer.go:
```diff
@@ -875,6 +937,44 @@ func (gce *GCECloud) firewallNeedsUpdate(name, serviceName, region, ipAddress st
+		return fmt.Errorf("error getting cluster ID: %v", err)
+	}
+	fwNameSuffix := MakeHealthCheckFirewallNameSuffix(clusterId, hcName, isNodesHealthCheck)
+	fwName := makeFirewallName(fwNameSuffix)
```
Regarding the use of `k8s-fw`... this is actually referring to the forwarding rule for an ingress obj. 

From our discussions with Tim, I think the desire is to get rid of these segments. Ideally, clusterId is always the second field of the string. `k8s-clusterid-....`
Would it be possible to adopt the naming scheme that I'm using?  I haven't looked at your other names just yet.
https://github.com/nicksardo/kubernetes/blob/gke-ilb/pkg/cloudprovider/providers/gce/gce_loadbalancer_names.go
Any comments on it?

MrHohn replied:
Make sense to follow the same naming pattern. I'm using `k8s-{clusterid}-node-hc` for the http healthcheck resource (for healthchecking nodes), we can use it for the corresponding firewall as well.

Since we didn't create firewall for nodeport healthcheck before, following the proposed schema ( `k8s-{clusterid}-{namespace}-{name}-{shortid}-hc`) also seems right. Though I'm not sure should we migrate the old http healthcheck resource (for healthchecking nodeport) to use the new name or should we leave it as is (for now it is something like` aXXXXXXX`)?

nicksardo replied:
> I'm using k8s-{clusterid}-node-hc for the http healthcheck resource (for healthchecking nodes), we can use it for the corresponding firewall as well.

Great, thanks

> Though I'm not sure should we migrate the old http healthcheck resource (for healthchecking nodeport) to use the new name or should we leave it as is (for now it is something likeaXXXXXXX)?

Since it's a different type of HC, I'm not concerned about using the same naming scheme. Migrating names is too much of a burden right now anyway.

MrHohn replied:
I think it makes sense to let existing resources (http healthcheck for nodeport, targetpool, forwarding rule, firewall rule for ELB svc) keep the old name now, and we can deal with the renaming/migration later if needed?

MrHohn replied:
> Since it's a different type of HC, I'm not concerned about using the same naming scheme. Migrating names is too much of a burden right now anyway.

Thanks, that is what I thought :+1:

nicksardo replied:
Yes, definitely.

MrHohn replied:
Regarding the name of firewall for nodeport healthcheck (`k8s-{clusterid}-{namespace}-{name}-{shortid}-hc`), not sure how to generate it since your commit is not in yet?

MrHohn replied:
Per discussion, I will use `"k8s-{hcName}-hc"` for now.

MrHohn replied:
Done.

nicksardo commented on pkg/cloudprovider/providers/gce/gce_loadbalancer.go:
```diff
@@ -875,6 +937,44 @@ func (gce *GCECloud) firewallNeedsUpdate(name, serviceName, region, ipAddress st
+	// Validate firewall fields.
+	if fw.Description != desc ||
+		!sameStringArray(fw.SourceRanges, sourceRanges.StringSlice()) ||
+		!sameStringArray(packProtocolsPortsFromFirewallAlloweds(fw.Allowed), packProtocolsPortsFromServicePorts(ports)) {
```
I think these helper funcs (`packProtocolsPortsFromFirewallAlloweds` and `packProtocolsPortsFromServicePorts`) are nice, but overkill and may cause an error eventually.  You're making them easy to compare; however, they are actually different. ServicePorts can contain multiple ports, one with TCP and one with UDP.  The load balancer is assembled using the first protocol, so the other ports' protocols don't matter.  The `firewallObject` func performs this logic, but you're not doing that on the other side of the equals.  It's currently working because the health check only requires one port with one protocol.

Since these helpers are only used here, I'd suggest going with simpler/smaller logic that matches current implementation in gce_loadbalancer. I also used the same logic in ILB.
https://github.com/nicksardo/kubernetes/blob/a6fe71bd6c7746ddd21fa008420d1c7ddac69698/pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go#L68 
Thoughts?

MrHohn replied:
Good catch! I missed that part. Sounds good to make it simpler.

MrHohn replied:
Done.

nicksardo commented on pkg/cloudprovider/providers/gce/gce_firewall.go:
```diff
@@ -112,3 +115,33 @@ func (gce *GCECloud) UpdateFirewall(name, desc string, sourceRanges netsets.IPNe
+	res := sets.NewString(result...)
+	exp := sets.NewString(expected...)
+	return res.Equal(exp)
+}
```
Could we move this func to gce_utils?

nicksardo replied:
Also, rename to `equalStringSets`

MrHohn replied:
Found we already have `slicesEqual()` in gce_loadbalancer.go. Removing this function and moving that to gce_utils.

MrHohn replied:
Done.

nicksardo commented on pkg/cloudprovider/providers/gce/gce_loadbalancer.go:
```diff
@@ -447,14 +452,25 @@ func (gce *GCECloud) EnsureLoadBalancerDeleted(clusterName string, service *v1.S
 		gce.region)
 
-	var hc *compute.HttpHealthCheck
+	hcNames := []string{}
```
Nit: In every employer of mine that used Go, this way was more preferable: 
`var hcNames []string`

`append` handles the initialization.

MrHohn replied:
Done.

nicksardo commented on pkg/cloudprovider/providers/gce/gce_loadbalancer.go:
```diff
@@ -509,15 +525,15 @@ func (gce *GCECloud) deleteForwardingRule(name, region string) error {
 }
 
-func (gce *GCECloud) deleteTargetPool(name, region string, hc *compute.HttpHealthCheck) error {
+func (gce *GCECloud) deleteTargetPool(name, region string, hcNames ...string) error {
```
Just to write down what we talked about offline, we could delete the health check that's linked from the targetpool. I suppose it's trading one DELETE (one of the health checks) call for one GET call (target pool). I don't know if it buys us anything other than possible code cleanliness. 

Just an idea for you to explore - up to you on whether we go with it.

MrHohn replied:
Will update this function when _ServiceLister_ arrives.

nicksardo commented on pkg/cloudprovider/providers/gce/gce_loadbalancer.go:
```diff
@@ -875,18 +938,46 @@ func (gce *GCECloud) firewallNeedsUpdate(name, serviceName, region, ipAddress st
-		return false
+func (gce *GCECloud) ensureHttpHealthCheckFirewall(serviceName, ipAddress, region string, hosts []*gceInstance, hcName string, hcPort int32, isNodesHealthCheck bool) error {
+	// Prepare the firewall params for creating / checking.
+	desc := `{"kubernetes.io/loadbalancer-nodes-healthcheck"}`
```
This isn't valid JSON. It needs to be a string map to match current descriptions.

I don't think we need to specify that it's "loadbalancer-nodes-healthcheck" in the description. It's already known to be a health check.  We could reiterate the cluster id? `{"kubernetes.io/cluster-id":"%s"}` Or cluster name? `gke-mycluster-123123` Thoughts?

MrHohn replied:
Thanks, sounds good to put cluster id in it.

MrHohn replied:
Done.

nicksardo commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -90,6 +91,7 @@ type GCECloud struct {
 	nodeInstancePrefix       string   // If non-"", an advisory prefix for all nodes in the cluster
 	useMetadataServer        bool
 	operationPollRateLimiter flowcontrol.RateLimiter
+	nodesHealthCheckLock     sync.Mutex
```
As discussed offline, we should rename this and add a large comment.

MrHohn replied:
Done.

thockin commented:
/approve

MrHohn commented:
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/45524/pull-kubernetes-node-e2e/31916/

MrHohn commented:
@k8s-bot pull-kubernetes-node-e2e test this

MrHohn commented:
/assign @bowei

MrHohn commented:
@bowei Would it be better if you have bandwidth to take a quick look?

MrHohn commented:
@nicksardo Renamed Id to ID. Let me know if that is not what you meant.

nicksardo commented:
Perfect, thank you.

bowei commented on pkg/cloudprovider/providers/gce/gce_util.go:
```diff
@@ -115,3 +125,17 @@ func splitProviderID(providerID string) (project, zone, instance string, err err
 	}
 	return matches[1], matches[2], matches[3], nil
 }
+
```
Don't we have this as part of the `set` type already?

MrHohn replied:
Changed to use the _sets_ package.

bowei commented on test/e2e/firewall.go:
```diff
@@ -45,42 +46,71 @@ var _ = framework.KubeDescribe("Firewall rule", func() {
 		firewallTestSourceRanges := []string{"0.0.0.0/1", "128.0.0.0/1"}
 		serviceName := "firewall-test-loadbalancer"
 
+		By("Fetching cluster ID")
```
[minor] Fetch -> Get, the method is called `GetClusterID`

MrHohn replied:
Done.

bowei commented on test/e2e/framework/firewall_util.go:
```diff
@@ -68,6 +71,32 @@ func ConstructFirewallForLBService(svc *v1.Service, nodesTags []string) *compute
+}
+
+// ConstructHealthCheckFirewallForLBService returns the expected GCE firewall rule for a loadbalancer type service
+func ConstructHealthCheckFirewallForLBService(clusterID string, svc *v1.Service, nodesTags []string, isNodesHealthCheck bool) *compute.Firewall {
```
is this duplicated code from the cloud provider?

MrHohn replied:
A similar function in the GCE cloud provider is gce.firewallObject(), but that bakes in too many cloud provider specific details and is sort of heavy.

bowei commented on test/e2e/framework/firewall_util.go:
```diff
@@ -325,3 +357,39 @@ func VerifyFirewallRule(res, exp *compute.Firewall, network string, portsSubset
+
+	condition := func() (bool, error) {
+		fw, err = gceCloud.GetFirewall(fwName)
+		if err != nil && exist ||
```
check that the return is notfound for !exist case?

MrHohn replied:
Good catch! Added a check.

bowei commented on test/e2e/framework/firewall_util.go:
```diff
@@ -325,3 +357,39 @@ func VerifyFirewallRule(res, exp *compute.Firewall, network string, portsSubset
 	return nil
 }
+
+func WaitForFirewallRuleExistOrNot(gceCloud *gcecloud.GCECloud, fwName string, exist bool, timeout time.Duration) (*compute.Firewall, error) {
```
shorter name WaitForFirewallRule or WaitForFirewallRuleStatus

MrHohn replied:
Done.

bowei commented on test/e2e/framework/firewall_util.go:
```diff
@@ -325,3 +357,39 @@ func VerifyFirewallRule(res, exp *compute.Firewall, network string, portsSubset
+	if !clusterIDExists {
+		return "", fmt.Errorf("cluster ID not set")
+	}
+	if providerIDExists && providerID != clusterID {
```
providerID != clusterID seems redundant?

MrHohn replied:
@nicksardo I copied it from ClusterID.GetID(), feel like it does seem redundant?

nicksardo replied:
At the start of a new cluster, both IDs are equal. We only know that they're federated when we know they aren't equal.  However, providerId was added later, so it's possible it's not set at all for old clusters.

MrHohn replied:
Removed the redundant check both here and in gce_clusterid.go.

bowei commented on pkg/cloudprovider/providers/gce/gce_healthchecks.go:
```diff
@@ -17,11 +17,31 @@ limitations under the License.
+)
+
+func init() {
+	flag.IntVar(&lbNodesHealthCheckPortFlag, "cloud-provider-gce-lb-nodes-healthcheck-port", ports.ProxyHealthzPort, "Port opened on nodes for LB health checks")
```
Seems like the wrong place for this flag, as it is not a property of GCE. Is it not possible to pass in the port when the health check is created?

MrHohn replied:
Removed the flag.

MrHohn commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -90,6 +91,10 @@ type GCECloud struct {
 	nodeInstancePrefix       string   // If non-"", an advisory prefix for all nodes in the cluster
 	useMetadataServer        bool
 	operationPollRateLimiter flowcontrol.RateLimiter
+	// sharedResourceLock is used to allow the controller to assemble a full
```
Thanks! Applied.

MrHohn commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -233,14 +238,15 @@ func CreateGCECloud(projectID, region, zone string, managedZones []string, netwo
 		nodeInstancePrefix:       nodeInstancePrefix,
 		useMetadataServer:        useMetadataServer,
 		operationPollRateLimiter: operationPollRateLimiter,
+		sharedResourceLock:       sync.Mutex{},
```
Actually none of above variables is guarded by this lock, so I kept it here.

bowei replied:
What is the purpose of this lock then? Can you document it?

MrHohn replied:
Humm, it is already documented in the GCECloud struct above?

bowei replied:
no need to initialize this

MrHohn replied:
Done.

MrHohn committed:
Create nodes health checks for non-OnlyLocal services

MrHohn committed:
Update firewall e2e test for LB healthcheck firewall

MrHohn commented:
Addressed comments and squashed commits. Unsquashed commits could be found on https://github.com/MrHohn/kubernetes/commits/l4-lb-healthcheck-unsquashed.

Serial and slow e2e tests passed on my cluster.

MrHohn commented:
@k8s-bot pull-kubernetes-verify test this

bowei commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -80,7 +81,7 @@ type GCECloud struct {
 	containerService         *container.Service
 	clientBuilder            controller.ControllerClientBuilder
-	ClusterId                ClusterId
+	ClusterID                ClusterID
```
[minor] move to top in its own block as this is public

bowei commented on pkg/cloudprovider/providers/gce/gce_healthchecks.go:
```diff
@@ -178,3 +190,59 @@ func (gce *GCECloud) ListHealthChecks() (*compute.HealthCheckList, error) {
+// balancers (l4) for performing health checks.
+func MakeHealthCheckFirewallName(clusterID, hcName string, isNodesHealthCheck bool) string {
+	// TODO: Change below fwName to match the proposed schema: k8s-{clusteriD}-{namespace}-{name}-{shortid}-hc.
+	fwName := "k8s-" + hcName + "-http-hc"
```
why not 

```
if isNodesHealthCheck {
  return makeNodesHealthCheckName(clusterID) + "-http-hc"
}
return "k8s-" + hcName + "-http-hc"
```

bowei commented on pkg/cloudprovider/providers/gce/gce_healthchecks.go:
```diff
@@ -17,11 +17,23 @@ limitations under the License.
 )
 
+const (
+	minNodesHealthCheckVersion = "1.7.0"
```
make this a var of type version and use init() to initialize (no need to parse it every time in the routines below)

bowei commented on pkg/cloudprovider/providers/gce/gce_healthchecks.go:
```diff
@@ -178,3 +190,59 @@ func (gce *GCECloud) ListHealthChecks() (*compute.HealthCheckList, error) {
+// isAtLeastMinNodesHealthCheckVersion checks if a version is higher than
+// `minNodesHealthCheckVersion`.
+func isAtLeastMinNodesHealthCheckVersion(vstring string) bool {
+	minVersion, err := utilversion.ParseGeneric(minNodesHealthCheckVersion)
```
see comment above about making the parsed copy a var

bowei commented on pkg/cloudprovider/providers/gce/gce_util.go:
```diff
@@ -115,3 +125,12 @@ func splitProviderID(providerID string) (project, zone, instance string, err err
 }
+
+func equalStringSets(x, y []string) bool {
+	if len(x) != len(y) {
```
seems redundant...

bowei reviewed (commented):
some nits

MrHohn commented:
@nicksardo says he will fix the nits in his PR, thanks!

bowei commented:
/lgtm

MrHohn committed:
Autogenerated files

MrHohn commented:
@nicksardo Rebased.

MrHohn commented:
@k8s-bot pull-kubernetes-node-e2e test this

freehan commented:
/lgtm

MrHohn commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this

alok87 commented:
@MrHohn Do we have health check for AWS ELBs? We want to keep adding/removing nodes in the ELB based on health. What is the right way to do it.
